### ISSUE TITLE
Different values on `Circuit.YCurrents()`

### ISSUE THREAD

**Issue by felipemarkson** [dss-extensions/OpenDSSDirect.jl#86]
Hi!

Maybe I'm missing something, but I was making a test and found a difference between currents get from `Circuit.YCurrents()` and what they should be. Here is an example:

**DSS File**
Just a 3-bus 13.8kV balanced system with one 1MW/0.5MVAr 3ph load.
```
// main.dss
new   circuit.test  basekv=13.8 bus1=SOURCEBUS   pu=1.0 MVAsc3=999999 MVAsc1=999999

New   Line.Line_1   bus1=SOURCEBUS   bus2=BUS_1 R1=3.0 X1=3.0 Length=1.0

New   Line.Line_2   bus1=BUS_1   bus2=BUS_2 R1=3.0 X1=3.0 Length=1.0

New   Load.Load_1   phases=3   conn=Wye   bus1=BUS_2   kw=1000   kvar=500   model=1 kV=13.8

Set Voltagebases=[13.8]
```

**The Julia file**

```Julia
using OpenDSSDirect
using Test
dss("""
    clear
    redirect main.dss
    solve   
""")
nodes = Circuit.YNodeOrder()
I = Circuit.YCurrents()
V = Circuit.YNodeVArray()
S = V.*conj.(I)

println(nodes[end]) # BUS_2.3
# Power injected by the load in the node BUS_2.3
P_MW = real(S[end])/1e6 
Q_MVAr = imag(S[end])/1e6

# The system nearly balanced, so we can just multiply by 3.
P_load = P_MW*3
Q_load = Q_MVAr*3

# So, the load calculated must have the same power as the .dss file
@test P_load ≈ -1.0 atol = 0.1 # This causes a error because P_load = -0.09716 not -1.0
@test Q_load ≈ -0.5 atol = 0.1 # This causes a error because P_load = -0.04858 not -0.5
```

Once run this code, the power injected is not the same as the load. Actually, you "need" to multiply the current by 10 to get the same values of the calculated.

Am I missing something? If yes, why the value of `Circuit.YCurrents()` is 10 times less than it should be?

**Comment by PMeira** [comment]
> Am I missing something?

It'd be too much to write in a comment here. Basically part of the current is represented in the Ymatrix (the "linear" part), part as a compensation current. [Quoting](https://sourceforge.net/p/electricdss/code/HEAD/tree/trunk/Version8/Distrib/Doc/OpenDSS%20Solution%20Technique.pdf?format=raw): 
> The nonlinear characteristics of the load, if any, are represented by the compensation current sources in the Norton equivalent.

The manual has some material, [this is a bit more targeted](https://sourceforge.net/p/electricdss/discussion/861976/thread/0321886436/af52/attachment/OpenDSS%20Power%20Flow%20Solution%20Method-2.pdf) and there is this paper too somewhere, [I believe this one](https://ieeexplore.ieee.org/document/6039829) -- quoting it:

> The compensation current is the difference between the current drawn by the nonlinear power conversion element and the linear portion of the element, if any, that is embedded in the system Y matrix.

If you have more specific questions about the methods, it might be better to ask on the official forum instead. For PQ loads, specifically, the code is not too complicated for rated operation: https://github.com/dss-extensions/dss_capi/blob/0bd6e3d92a8e922f4ce700117aa4e0a64f66c827/src/PCElements/Load.pas#L1201-L1235

This may help reading this code snippet without going through the whole codebase: https://sourceforge.net/p/electricdss/code/HEAD/tree/trunk/Version8/Distrib/Doc/PCElementEssentials.pdf?format=raw

**Comment by PMeira** [comment]
Maybe a simpler way to get an idea of how it works: change the load model to constant impedance and recheck the values you calculated. Then activate the load and look into `CktElement.Powers`, `.Currents` and `.Voltages`. If you want to go further, you could use `Yprim` and `NodeRef` and calculate the three previous vectors by yourself.

**Comment by felipemarkson** [comment]
Thanks for this answer @PMeira!

Just another question, there's a way, using the API, to get the Current injected in each node?

**Comment by PMeira** [comment]
Do you want to redo this? https://github.com/dss-extensions/dss_capi/blob/0bd6e3d92a8e922f4ce700117aa4e0a64f66c827/src/PCElements/PCElement.pas#L107

`Currents` there is what `YCurrents` returns: https://github.com/dss-extensions/dss_capi/blob/0bd6e3d92a8e922f4ce700117aa4e0a64f66c827/src/CAPI/CAPI_Circuit.pas#L1032

**Comment by felipemarkson** [comment]
Sorry, but I don't understand your answer.

What I mean is something like `Bus.Currents()::Vector{ComplexF64}` where each value in the `Vector` is the current injected in the node respectively.

**Comment by PMeira** [comment]
Per the method, `YCurrents` is already the vector of injections, so your previous question was a little odd.

From what I understand, you want `CktElement.Currents` instead. Like in the first snippet I linked in the my previous post, use `CktElement.NodeRef` directly (you probably can ignore values where `NodeRef[index] = 0`) as it provides an index into the voltage/current vectors for each node of an element. So, try just looping through the elements you want and sum away.

If you still don't fully understand the power flow method, I'd recommend simplifying your circuit even further (use a single line and a single phase) and then trying to implement it manually in a separate script.

And again, besides the docs, check the forum, there is more than a decade of questions there. Some more links that could help:
- https://sourceforge.net/p/electricdss/discussion/861976/thread/44c3412b/
- https://sourceforge.net/p/electricdss/discussion/861976/thread/92e3458d/

**Comment by felipemarkson** [comment]
Ok! Thanks, @PMeira!
